### PR TITLE
Assume Ingress identity also for cluster internal traffic via Ingress

### LINKED ISSUE #24536: CFP: Fix Ingress traffic interactions with other Policy enforcement
## Background

Currently, Cilium's Ingress support has implications when interacting with other Policies that are a bit surprising and aren't well explained.

When traffic arrives at the per-node Envoy from _outside_ the cluster (in Cilium identity speak, it has a _world_ identity) to a listener created by Cilium's Ingress (or Gateway API) support, it picks up the special _ingress_ identity during its routing through Envoy. This identity can then interact with Network Policy (whether they are CiliumNetworkPolicy or ClusterWideCiliumNetworkPolicy) by being used in an ingress or egress policy.

> **Note**
> Yes, the fact that traffic coming into the cluster is called ingress, as well as traffic going _to_ an identity is confusing. This term overloading has come from two different directions and is tricky to fix. In this issue, `Ingress` means the Kubernetes resource `Ingress`, and other uses will be disambiguated with a modifier like "ingress identity", "ingress policy", or "ingress traffic".

However, when traffic arrives at the per-node Envoy from _inside_ the cluster, _and_ is bound for an ingress listener, the original security identity is maintained. This is because Cilium's Ingress support is built on our Layer 7 Policy support, and uses the same Envoys. This means that if users wish to have _in-cluster_ workloads talk to the "outside" of their Ingress or Gateway API listeners _and_ have Network Policy that restricts traffic between workloads, they may need to allow an unintendedly large amount of traffic in a Network Policy.

This has been the cause of some confusion and outages with a number of Cilium users, and the number will only grow as more people use Cilium's Ingress and Gateway API support.

In an old-school, physical network, that had a dedicated firewall device listening on certain open ports on an internet connection, this type of connection was called "hairpinning" (like a hairpin turn on a racetrack), and had a series of special options.

In Cilium's case, fully supporting Ingress hairpinning could safely have internal traffic pick up the _ingress_ identity, because the Ingress and Gateway API listeners are effectively performing some Policy enforcement themselves. This is because both Ingress and Gateway API will only route _requested_ traffic to their defined backends - they are effectively an allowlist for very specific Layer 7 traffic when passing through the per-node Envoys.

## Proposed changes

This CFP proposes two changes:
- [ ] Firstly, _all_ traffic passing through an Ingress or Gateway API listener should assume the _ingress_ identity. This includes traffic originating inside the cluster, and will mean that traffic from inside will _lose_ its internal identity. This is hairpinning for internal traffic.
- [ ] Secondly, the documentation will be updated to call out this interaction, and make it clear that by having traffic originating from inside the cluster go to an Ingress listener, it's _logically_ "leaving" the cluster and becoming _ingress_ identity traffic instead.

### PR DESCRIPTION
**Pull request by jrajahalme**
Add 'useOriginalSourceAddr' parameter to ParseResources() to inform if
the listener should use original source address or not. Pass this as
'true' from C/CEC watchers if OwnerReferences does not contain
Kind=Ingress. That is, do not use original source addressing for Ingress
listeners. This makes upstream connections from Ingress listeners to use
node's allocated Ingress addresses in all cases, so that the source if
Ingress traffic within the cluster is seen as 'ReservedIdentityIngress'
(8).

Fixes: #24536

```release-note
Assume Ingress identity for cluster internal traffic through Cilium Ingress for policy enforcement.
```

### CODE REVIEW

**Comment by jrajahalme** [comment]
Updated to new `cilium/proxy` build.

**Comment by jrajahalme** [comment]
IntegrationTest failed to start due to:
```
gpg: Can't check signature: No public key
ERROR: Failed to verify clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz
```
restarted.

**Comment by jrajahalme** [comment]
Updated to pick ownerReference change in CEC for shared Ingress.

**Comment by jrajahalme** [comment]
@sayboras Opened for reviews, but blocking merge until https://github.com/cilium/proxy/pull/167 is merged.

**Comment by sayboras** [comment]
>  Opened for reviews, but blocking merge until [cilium/proxy#167](https://github.com/cilium/proxy/pull/167) is merged.

Thanks, the above PR is merged, the changes look good to me, so I think it's just the matter of dropping debug commit, and update image from latest cilium/proxy build.

**Comment by sayboras** [comment]
Perform the above two items and kick off the tests.

**Comment by jrajahalme** [comment]
Rebased to merged `cilium/proxy` image and to main.

**Comment by jrajahalme** [comment]
Interesting but likely unrelated panic in runtime tests (https://jenkins.cilium.io/job/Cilium-PR-Runtime-net-next/6152/testReport/(root)/Suite-runtime/RuntimeDatapathPrivilegedUnitTests_Run_Tests/):
```
	 [16:25:27] === RUN   Test_httpRouteReconciler_Reconcile/http_route_with_non-matching_hostname
	 [16:25:27] level=info msg="Reconciling HTTPRoute" controller=httpRoute resource=default/http-route-with-non-matching-hostname subsys=gateway-controller
	 [16:25:27] level=info msg="Successfully reco
...[truncated 1067983 chars]...
 [16:36:22] 	/usr/local/go/src/testing/testing.go:1576 +0x10b                                                                                             
	 [16:36:22] created by testing.(*T).Run                                                                                                                  
	 [16:36:22] 	/usr/local/go/src/testing/testing.go:1629 +0x3ea                                                                                             
```

**Comment by jrajahalme** [comment]
/test-1.25-5.4

**Comment by jrajahalme** [comment]
/test-1.26-4.19

**Comment by jrajahalme** [comment]
Restarted tests that failed to start.

**Comment by jrajahalme** [comment]
test-1.25-5.4 failed on a known flake #24966, restarting

**Comment by jrajahalme** [comment]
/test-1.25-5.4

**Comment by jrajahalme** [comment]
test-1.27-net-next had not started, starting manually.

**Comment by jrajahalme** [comment]
/test-1.27-net-next

**Comment by jrajahalme** [comment]
> it's just a matter of dropping debug commit

@sayboras Maybe we should keep the debug logging for the Ingress and Gateway-API test runs? Could come in handy when something breaks?

**Comment by sayboras** [comment]
> Maybe we should keep the debug logging for the Ingress and Gateway-API test runs? Could come in handy when something breaks?

Yup, we can enable it separately to avoid CI re-run in this PR, or in case you make another update here.

**Comment by sayboras** [comment]
/test-1.27-net-next

**Comment by dylandreimerink** [comment]
FYI. CI is failing on a test added yesterday https://github.com/cilium/cilium/pull/24835, so this branch might need to be rebased to get it to go away. (hitting the same on one of my PRs)

**Review by tklauser — APPROVED** [review]
Looks good for @cilium/vendor changes.

**Comment by jrajahalme** [comment]
Rebased to pick up an egress gw fix, this should be good to review/merge given that @sayboras added an integration test for this in https://github.com/cilium/cilium-cli/pull/1533.

File: pkg/k8s/watchers/cilium_envoy_config.go [hunk 1]
```diff
@@ -83,6 +84,16 @@ func (k *K8sWatcher) ciliumEnvoyConfigInit(ctx context.Context, ciliumNPClient c
 	k.k8sAPIGroups.AddAPI(k8sAPIGroupCiliumEnvoyConfigV2)
 }
 
+// isIngressKind returns true if any of the OwnerReferences is Kind "Ingress"
+func isIngressKind(meta *meta_v1.ObjectMeta) bool {
+	for _, owner := range meta.OwnerReferences {
+		if owner.Kind == "Ingress" {
+			return true
+		}
+	}
```
Comment (sayboras) [inline pkg/k8s/watchers/cilium_envoy_config.go:93]: the same might need to be done for GatewayAPI resources (e.g. Gateway)

**Review by sayboras — APPROVED** [review]
LGTM, one small comment as per below.

**Review by christarazi — APPROVED** [review]

File: pkg/k8s/watchers/cilium_envoy_config.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/k8s/watchers/cilium_envoy_config.go:93]: Let's do that on a separate PR as that would not need to be backported to 1.13, right?

**Comment by michi-covalent** [comment]
why doesn't this pull request have `backport-pending/1.13` label? there is https://github.com/cilium/cilium/pull/25019 open for backporting this to v1.13 branch 👀

**Comment by jrajahalme** [comment]
> why doesn't this pull request have `backport-pending/1.13` label? there is #25019 open for backporting this to v1.13 branch 👀

Fixed that, sorry. This was an author-backport and so far for me that has been a manual process. Forgot to flip the label, that's all.

File: pkg/k8s/watchers/cilium_envoy_config.go (hunk 1 above)
Comment (rauanmayemir) [reply pkg/k8s/watchers/cilium_envoy_config.go:93]: @jrajahalme pinging you directly for exposure in #28254. it seems this hasn't been backported for gateway api?

**Pull request merged**